Sources/MuxUploadSDK/Internal Utilities/Reporter.swift

```diff
@@ -0,0 +1,61 @@
+//
+//  Reporter.swift
+//  
+//
+//  Created by Liam Lindner on 3/16/23.
+//
+
+import Foundation
+
+class Reporter: NSObject {
+    var session: URLSession?
+    var pendingUploadEvent: UploadEvent?
+
+    // MARK: - Constructor -
+    static let sharedInstance = Reporter()
+
+    fileprivate override init() {
+        super.init()
+
+        let sessionConfig: URLSessionConfiguration = URLSessionConfiguration.default
+        session = URLSession(configuration: sessionConfig, delegate: self, delegateQueue: nil)
+    }
+
+    func report(start_time: TimeInterval, end_time: TimeInterval, file_size: UInt64, video_duration: Double) -> Void {
+        self.pendingUploadEvent = UploadEvent(start_time: start_time, end_time: end_time, file_size: file_size, video_duration: video_duration)
+
+        let request = self.generateRequest(url: URL(string: "https://mobile.muxanalytics.com")!)
+
+        let dataTask = session?.dataTask(with: request as URLRequest, completionHandler: { (data, response, error) -> Void in
+            self.pendingUploadEvent = nil
+        })
+        dataTask?.resume()
+    }
+
+    private func generateRequest(url: URL) -> URLRequest {
+        let request = NSMutableURLRequest(url: url,
+                                          cachePolicy: .useProtocolCachePolicy,
+                                          timeoutInterval: 10.0)
+        request.httpMethod = "POST"
+        request.setValue("application/json", forHTTPHeaderField: "Accept")
+        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
+
+        do {
+            let jsonData = try JSONEncoder().encode(self.pendingUploadEvent)
+            request.httpBody = jsonData
+        } catch _ as NSError {}
+
+        return request as URLRequest
+    }
+}
+
+extension Reporter: URLSessionDelegate, URLSessionTaskDelegate {
+    public func urlSession(_ session: URLSession, task: URLSessionTask, willPerformHTTPRedirection response: HTTPURLResponse, newRequest request: URLRequest, completionHandler: @escaping (URLRequest?) -> Swift.Void) {
+        if(self.pendingUploadEvent != nil) {
+            if let redirectUrl = request.url {
+                let request = self.generateRequest(url: redirectUrl)
+                completionHandler(request)
+            }
+        }
+    }
+}
```

Comment:
By default, the client does not send it's request data after it gets redirected. This extension is necessary to regenerate the request, otherwise the redirect would receive a GET with no body.

Reply:
Do we know why a redirect happens in this case?

Reply:
Yeah, it's because of how I configured the backend. Tray.io provides a webhook that receives POSTs. I didn't want to ship the SDK using the tray.io URL so I set up a load balancer with a mux domain in front of it and a redirection rule to send traffic to the Tray endpoint. Due to how HTTP works, if you redirect a POST using HTTP 301, the client will send a GET instead of a POST to the redirect URL.

This is a security feature so that POST data doesn't get sent to an unknown destination due to a redirect. Because we set up this redirection, I'm ok with the client sending the POST after the redirection.

Reply:
Yeah redirecting to tray.io via our own domain makes sense! 

Out of curiosity can the load balancer send a `307` for this? From my skim of the [MDN HTTP reference](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/307) `307` fits closer to our case: send the post request to our own domain and then redirect it elsewhere that may change in future. A `301` is indicates permanent redirect.

Since `URLSession` is a high-level networking API and has its own internal logic to handle these status codes I'm concerned a `301` may lead the SDK to cache the `tray.io` URL and attempt to use it even after we e.g. adjust the load balancer.

Feel free to take this as a follow-up in a separate PR because I don't think this a major concern, just a corner case to consider.

Reply:
Unfortunately, there's no way to send a 307 from the load balancer. Only 301 or 302 are supported https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/lb_listener#redirect

I hadn't thought about the possibility of the tray.io url getting cached, but I see what you mean. What do you think if we sent a 302 instead so it's considered a temporary redirect. Do you think that would be less likely to get cached?